numpyro/infer/elbo.py

```diff
@@ -280,3 +282,266 @@ def single_particle_elbo(rng_key):
         return -(stop_gradient(renyi_elbo - weighted_elbo) + weighted_elbo)
+
+
+def _get_plate_stacks(trace):
+    """
+    This builds a dict mapping site name to a set of plate stacks. Each
+    plate stack is a list of :class:`CondIndepStackFrame`s corresponding to
+    an :class:`plate`. This information is used by :class:`Trace_ELBO` and
+    :class:`TraceGraph_ELBO`.
+    """
+    return {
+        name: [f for f in node["cond_indep_stack"]]
+        for name, node in trace.items()
+        if node["type"] == "sample"
+    }
+
+
+class MultiFrameTensor(dict):
+    """
+    A container for sums of Tensors among different :class:`plate` contexts.
+    Used in :class:`~numpyro.infer.elbo.TraceGraph_ELBO` to simplify
+    downstream cost computation logic.
+
+    Example::
+
+        downstream_cost = MultiFrameTensor()
+        for site in downstream_nodes:
+            downstream_cost.add((site["cond_indep_stack"], site["log_prob"]))
+        downstream_cost.add(*other_costs.items())  # add in bulk
+        summed = downstream_cost.sum_to(target_site["cond_indep_stack"])
+    """
+
+    def __init__(self, *items):
+        super().__init__()
+        self.add(*items)
+
+    def add(self, *items):
+        """
+        Add a collection of (cond_indep_stack, tensor) pairs. Keys are
+        ``cond_indep_stack``s, i.e. tuples of :class:`CondIndepStackFrame`s.
+        Values are :class:`torch.Tensor`s.
+        """
+        for cond_indep_stack, value in items:
+            frames = frozenset(f for f in cond_indep_stack)
+            assert all(f.dim < 0 and -jnp.ndim(value) <= f.dim for f in frames)
+            if frames in self:
+                self[frames] = self[frames] + value
+            else:
+                self[frames] = value
+
+    def sum_to(self, target_frames):
+        total = None
+        for frames, value in self.items():
+            for f in frames:
+                if f not in target_frames and jnp.shape(value)[f.dim] != 1:
+                    value = value.sum(f.dim, keepdims=True)
+            while jnp.shape(value) and jnp.shape(value)[0] == 1:
+                value = value.squeeze(0)
+            total = value if total is None else total + value
+        return 0.0 if total is None else total
+
+    def __repr__(self):
+        return "%s(%s)" % (
+            type(self).__name__,
+            ",\n\t".join(["({}, ...)".format(frames) for frames in self]),
+        )
+
+
+def _identify_dense_edges(trace):
+    succ = {}
+    for name, node in trace.items():
+        if node["type"] == "sample":
+            succ[name] = set()
+    for name, node in trace.items():
+        if node["type"] == "sample":
+            for past_name, past_node in trace.items():
+                if past_node["type"] == "sample":
+                    if past_name == name:
+                        break
+                    # XXX: different from Pyro, we always add edge past_name -> name
```

Comment:
This is the only difference comparing to [Pyro implementation](https://github.com/pyro-ppl/pyro/blob/4a61ef2f9050ef81d1b0aa148d14ecc876f24a51/pyro/poutine/trace_messenger.py#L26).

Reply:
can you please clarify?

Reply:
I don't actually understand what's the role of `counter` in [CondIndepStackFrame](https://github.com/pyro-ppl/pyro/blob/4a61ef2f9050ef81d1b0aa148d14ecc876f24a51/pyro/poutine/indep_messenger.py#L15). I just guess that it is used for `iplate` or something like that. I assume `query.counter == target.counter == 0`, so the `if` branch in Pyro implementation does not trigger. Do I understand correctly?

Reply:
yes counter is only for non-vectorized plates. i see you're saying you don't need to accommodate within-non-vectorized-plate-independence